torch/csrc/autograd/python_variable.cpp

```diff
@@ -64,7 +65,12 @@ void concrete_decref_fn(const c10::impl::PyInterpreter* self, PyObject* pyobj) {
     return;
 
   pybind11::gil_scoped_acquire gil;
-  if (Py_REFCNT(pyobj) > 1) {
+  // Two possibilities:
+  // 1. We are decref-ing a tensor. Then we must be careful about
+  // PyObject resurrection (this only applies to Tensors, see THPVariable_clear).
+  // 2. We are decref-ing some other Python object. We don't do
+  // PyObject resurrection on non-Tensors, so we just carry on as usual
+  if (THPVariable_Check(pyobj) && Py_REFCNT(pyobj) > 1) {
```

Comment:
@albanD @ezyang I'm still investigating, but does the following ring a bell by any chance?

On an internal test, THPVariable_Check (which looks like the following)
```
inline bool THPVariable_Check(PyObject *obj)
{
  return THPVariableClass && PyObject_IsInstance(obj, THPVariableClass);
}
```
calls `PyObject_IsInstance` which returns -1. The actual error message looks like:
```
StopIteration

The above exception was the direct cause of the following exception:

SystemError: <built-in method __instancecheck__ of torch._C._TensorMeta object at 0x619001768890> returned a result with an error set
```

Reply:
Pretty much a coincidence. But maybe https://github.com/pytorch/pytorch/pull/64551 would help?

Reply:
You're probably choking on this code with a meta tensor:

```
// TODO: Deprecate this instancecheck entirely.  It's here to make
// instanceof(t, torch.FloatTensor) work, but we are not going to keep
// adding torch.QuantizedIntTensor classes for every new tensor type
// we add...
static PyObject* Tensor_instancecheck(PyObject* _self, PyObject* arg) {
  HANDLE_TH_ERRORS
  auto self = (PyTensorType*)_self;
  if (THPVariable_Check(arg)) {
    const auto& var = THPVariable_Unpack(arg);
    // NB: This is a little unfortunate, in that if I do an isinstance check
    // against torch.cuda.FloatTensor, this will immediately initialize CUDA.
    // I originally thought that it would not be possible for aten_type_ to
    // be nullptr if you had a tensor of some type, in which case you can
    // skip initializing aten_type(), but TestAutograd.test_type_conversions
    // seems to violate this property (for whatever reason.)
    //
    // TODO: Stop using legacyExtractDispatchKey here (probably need to build
    // in instanceof checking to Tensor class itself)
    if (legacyExtractDispatchKey(var.key_set()) == self->get_dispatch_key() &&
        var.scalar_type() == static_cast<ScalarType>(self->scalar_type)) {
      Py_RETURN_TRUE;
    }
  }
  Py_RETURN_FALSE;
  END_HANDLE_TH_ERRORS
}
```

Reply:
Thanks for the suggestions. I was able to get a very weird repro on some internal build (still working on getting something that doesn't depend on internal bits...):
```
@async_test
async def test_one_frame(self):
    net = nn.Linear(2, 1)

    def foo():
        x = torch.nn.Parameter(torch.randn(3))
        y = x * x
        del x
        return y

    y = foo()
```
It looks like part of the problem is creating a nn.Parameter, then flipping the ownership such that the Tensor owns the nn.Parameter, and finally calling `PyObject_IsInstance(pyobj, THPVariableClass)`. Furthermore it looks like the async nature of the test is required to get this to trigger. Not sure if there are any weird interactions with PyInterpreter and async.

Reply:
I'm pretty stumped on this, so I'm planning on attempting to investigating more and then going back to having a library split between TorchDispatchObject and PythonTorchDispatchObjectImpl tomorrow to avoid having to modify `concrete_decref_fn` to handle non-Tensor PyObjects.

Reply:
Are you still on an internal only repro? Give it a share and maybe someone else can take a look.

I think there is at least one bug related to PyObject preservation that I also got a difficult internal repro for, and I haven't be able to track it down. So maybe this is it.

Reply:
Yes, still on an internal-only repro (I can't figure out how to copy and paste the internal TestCase because it has a lot of dependencies). Let me cut it down to a really minimal case and then send it out.

Reply:
@ezyang @albanD Here's the internal repro if any of you two want to take a look: D30960775

Reply:
Was able to excise the internal dependencies to produce an OSS repro at https://github.com/pytorch/pytorch/pull/65068

Reply:
And this is now resolved! Thank you @albanD and @ezyang for helping to debug!